common/nexus/failure.go

```diff
@@ -47,65 +51,124 @@ func ProtoFailureToNexusFailure(failure *nexuspb.Failure) *nexus.Failure {
+// APIFailureToNexusFailure converts an API proto Failure to a Nexus SDK Failure setting the metadata "type" field to
+// the proto fullname of the temporal API Failure message.
 // Always returns a non-nil value.
-func APIFailureToNexusFailure(failure *failurepb.Failure) *nexus.Failure {
-	return &nexus.Failure{
-		Message: failure.GetMessage(),
+func APIFailureToNexusFailure(failure *failurepb.Failure) (nexus.Failure, error) {
+	// Unset message so it's not serialized in the details.
+	failureCopy := common.CloneProto(failure)
+	failureCopy.Message = ""
+	data, err := protojson.Marshal(failureCopy)
+	if err != nil {
+		return nexus.Failure{}, err
 	}
+	return nexus.Failure{
+		Message: failure.GetMessage(),
+		Metadata: map[string]string{
+			"type": failureTypeString,
+		},
+		Details: data,
+	}, nil
 }
 
-func UnsuccessfulOperationErrorToTemporalFailure(err *nexus.UnsuccessfulOperationError) *failurepb.Failure {
-	failure := &failurepb.Failure{
-		Message: err.Failure.Message,
-	}
-	if err.State == nexus.OperationStateCanceled {
-		failure.FailureInfo = &failurepb.Failure_CanceledFailureInfo{
-			CanceledFailureInfo: &failurepb.CanceledFailureInfo{
-				Details: nexusFailureMetadataToPayloads(err.Failure),
-			},
+// NexusFailureToAPIFailure converts a Nexus Failure to an API proto Failure.
+// If the failure metadata "type" field is set to the fullname of the temporal API Failure message, the failure is
+// reconstructed using protojson.Unmarshal on the failure details field.
+func NexusFailureToAPIFailure(failure nexus.Failure, retryable bool) (*failurepb.Failure, error) {
+	apiFailure := &failurepb.Failure{}
+
+	if failure.Metadata != nil && failure.Metadata["type"] == failureTypeString {
+		if err := protojson.Unmarshal(failure.Details, apiFailure); err != nil {
+			return nil, err
 		}
 	} else {
-		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{
+		payloads, err := nexusFailureMetadataToPayloads(failure)
+		if err != nil {
+			return nil, err
+		}
+		apiFailure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{
 			ApplicationFailureInfo: &failurepb.ApplicationFailureInfo{
 				// Make up a type here, it's not part of the Nexus Failure spec.
-				Type:         "NexusOperationFailure",
-				Details:      nexusFailureMetadataToPayloads(err.Failure),
-				NonRetryable: true,
+				Type:         "NexusFailure",
+				Details:      payloads,
+				NonRetryable: !retryable,
 			},
 		}
 	}
-	return failure
+	// Ensure this always gets written.
+	apiFailure.Message = failure.Message
+	return apiFailure, nil
+}
+
+func UnsuccessfulOperationErrorToTemporalFailure(opErr *nexus.UnsuccessfulOperationError) (*failurepb.Failure, error) {
+	var nexusFailure nexus.Failure
+	failureErr, ok := opErr.Cause.(*nexus.FailureError)
+	if ok {
+		nexusFailure = failureErr.Failure
+	} else if opErr.Cause != nil {
+		nexusFailure = nexus.Failure{Message: opErr.Cause.Error()}
+	} else {
+		nexusFailure = nexus.Failure{Message: "canceled"}
```

Comment:
How are we ensuring that only canceled failure doesn't have the `Cause` field `nil` value?

Reply:
Don't understand the question. Can you clarify? Do you mean "Why"?

Reply:
This code assumes that if `opErr.Cause` is `nil`, then the failure is that the operation was canceled. How are ensuring that this is always the case? That is, what if later a different operation error is returned without the `Cause` field being populated?

Reply:
Ah you're right, this isn't right. TBH, Cause should never be nil but I'll fix this fallback just in case.